src/NonLinearSolver.f90

```diff
@@ -0,0 +1,429 @@
+
+  !> @brief Interface for the @c jacobian method of a @c NonLinearSolverInterface_Base
+  !>        extension
+  !> @param this the @c NonLinearSolverInterface_Base object
+  !> @param x the @c VectorType to evaluate the Jacobian at
+  !> @param J the @c MatrixType containing the evaluation of the Jacobian at @c x
+  !>
+  SUBROUTINE nonlinearsolver_jacobian_absintfc(this,x,J)
+    IMPORT :: NonLinearSolverInterface_Base,VectorType,MatrixType
+    CLASS(NonLinearSolverInterface_Base),INTENT(IN) :: this
+    CLASS(VectorType),INTENT(INOUT) :: x
+    CLASS(MatrixType),INTENT(INOUT) :: J
+  ENDSUBROUTINE nonlinearsolver_jacobian_absintfc
+
+  !> @brief Interface for the @c solve method of a @c NonLinearSolver extension
+  !> @param this the @c NonLinearSolver object
+  !> @param x the @c VectorType containing the initial guess for the solution
+  !>        of the root-finding procedure
+  !>
+  SUBROUTINE nonlinearsolver_solve_absintfc(this,x)
+    IMPORT :: NonLinearSolver_Base,VectorType
+    CLASS(NonLinearSolver_Base),INTENT(INOUT) :: this
+    CLASS(VectorType),INTENT(INOUT) :: x
+  ENDSUBROUTINE nonlinearsolver_solve_absintfc
+ENDINTERFACE
+
+!> Module name
+CHARACTER(LEN=*),PARAMETER :: modName='NonLinearSolver'
+
+!
+!===============================================================================
+CONTAINS
+!
+!-------------------------------------------------------------------------------
+!> @brief Defines the interface for clearing a @c NonLinearSolverInterface_Base
+!> @param this the object to clear
+!>
+SUBROUTINE clearInterface_Base(this)
+  CLASS(NonLinearSolverInterface_Base),INTENT(INOUT) :: this
+
+  !Does nothing.  This is implemented in case extensions of the interface implement
+  !it so we can clear the extensions of this class when clearing the solver itself.
+
+ENDSUBROUTINE clearInterface_Base
+!
+!-------------------------------------------------------------------------------
+!> @brief Defines the interface for initializing a @c NonLinearSolver_Base object
+!> @param this the object to initialize
+!> @param ce the computing environment
+!> @param f the @c NonLinearSolverInterface_Base object to use for function evals
+!> @param plist the parameter list to use for initialization
+!>
+!> This procedure takes ownership of @c f and clears the input pointer.  It does
+!> not take ownership of @c ce.
+!>
+SUBROUTINE init_NonLinearSolverBase(this,ce,f,plist)
+  CLASS(NonLinearSolver_Base),INTENT(INOUT) :: this
+  TYPE(FutilityComputingEnvironment),TARGET,INTENT(IN) :: ce
+  CLASS(NonLinearSolverInterface_Base),INTENT(INOUT),POINTER :: f
+  TYPE(ParamType),INTENT(IN) :: plist
+  !
+  REAL(SRK),ALLOCATABLE :: bounds(:)
+
+  REQUIRE(.NOT.this%isInit)
+  REQUIRE(ASSOCIATED(f))
+  REQUIRE(plist%has('NonLinearSolver -> n'))
+  REQUIRE(plist%has('NonLinearSolver -> method'))
+  REQUIRE(plist%has('NonLinearSolver -> tolerance'))
+  REQUIRE(plist%has('NonLinearSolver -> bounds'))
+  CALL plist%get('NonLinearSolver -> bounds',bounds)
```

Comment:
Should bounds be required?

Reply:
Your suggestion about the function object enforcing the bounds makes a lot of sense, so I'll do that instead.  That will nullify this comment.

Reply:
done